### PR TITLE
 #161967008 Implement token generation

### PR DESCRIPTION
**Pull request by gitau-kinuthia**
#### What does this PR do?
- upgrade packages in requirements.txt
- fix url endpoints
- fix DEFAULT_AUTHENTICATION_CLASSES bug in settings
- set up the testing environment using pytest
- add test for token in authentication requests payload
- update requirements.txt
- Add a default model in core to be inherited in the application
- add token encoding functionality
- consume  the encode_token method in the authentication serializers
- increase test coverage by adding more tests
- set up logging for easier debugging
- Add JWT token generation during registration and login

#### Description of Task to be completed?
- Have JWT token encoding and decoding functionality 
- Have the refresh token endpoint working
`POST /api/users/`
`POST /api/users/login/`
`POST /api/users/refresh/`

#### How should this be manually tested?
1. `git clone https://github.com/andela/ah-jumanji.git`
2. `cd ah-jumanji`
3. `git checkout ft-generate-jwt-tokens-161967008`
4.  Setup a `virtual environment` and activate it
5. `pip install -r requirements.txt`
6. `python manage.py makemigrations`
7. `python manage.py migrate`
8. `python manage.py runserver`

Now use postman to work with the endpoints or type `pytest` on your terminal to run tests.

#### Any background context you want to provide?
Users once registered via the registration endpoint, should receive a JWT to be required on all subsequent calls to all other endpoints that require authentication.

#### What are the relevant pivotal tracker stories?
[#161967008](https://www.pivotaltracker.com/story/show/161967008)
[#161966966](https://www.pivotaltracker.com/story/show/161966966)

##### Signup
screen shot 2018-11-30 at 09 44 03

##### Login
screen shot 2018-11-30 at 09 44 50

#### Checklist:
- [x] At least 2 people have reviewed my PR
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My PR has one commit.

### CODE REVIEW

**Comment by ghost** [comment]
1. I cloned and carried out the instructions to test, and I got an error as below:
image

Is this unique to me? @other_collaborators, please try this and let me know.

2. I think the `PR` is too wordy. Try trim down the amount of writing you have done. (especially in responding to the qn `What does this PR do?` I think a summary would be fine.)

3. On the instructions for testing, consider using a numbered list. For example:
```
1. git clone <link-to-repo> && git checkout <branch-name>
2. Create and activate a `virtualenv`
3. pip install -r requirements.txt 
....
```

**Comment by ghost** [comment]
`UPDATE` : I had to `python manage.py makemigrations && python manage.py migrate` in order for this to work. If this is required, add it to the list of instructions under`How can this manually be tested?`, or detail how a tester can avoid errors like the one I had above.

**Review by mashafrancis — COMMENTED** [review]
Great job on the implementation. Maybe the addition to have just as @dmithamo1 has said, to update the PR on how it should be tested with the step-by-step instruction. If numbered, the better.

**Review by EmmanuelBeja — COMMENTED** [review]
You don't seem to be using package `os` on line 5 of backends.py. It would be wise to remove it.

**Comment by ghost** [comment]
I noticed this `PR` implements multiple features/chores, each of which has its own story on the `PT` board.
For example:
1. It upgrades the project to `Django 2.`, for which a separate story exists ([#161966998](https://www.pivotaltracker.com/story/show/161966998))

2. It sets up the schema for `Swagger` documentation, for which a separate story exists ([#161967007](https://www.pivotaltracker.com/story/show/161967007))

and etc.

I think these should be done separately, as structured in the `PT` board.

**Comment by GransonO** [comment]
> I noticed this `PR` implements multiple features/chores, each of which has its own story on the `PT` board.
> For example:
> 
> 1. It upgrades the project to `Django 2.`, for which a separate story exists ([#161966998](https://www.pivotaltracker.com/story/show/161966998))
> 2. It sets up the schema for `Swagger` documentation, for which a separate story exists ([#161967007](https://www.pivotaltracker.com/story/show/161967007))
> 
> and etc.
> 
> I think these should be done separately, as structured in the `PT` board.

In addition to what `Mithamo` has highlighted, let us try and implement one feature for each PR so all stories are easily traced.

**Review by ghost — DISMISSED** [review]
Please look into the changes suggested in previous comments.
This `PR` will need to be trimmed, so that it `only implements features/chores from one PT story`, to maintain the recommended workflow.

**Comment by gitau-kinuthia** [comment]
@dmithamo1 @GransonO  I had asked the TTL about this and he had mentioned that I should do 1 PR. However, I am working on splitting it into two and will be raising the separate PRs. @EmmanuelBeja noted and removed.

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,63 @@
...
+    # @staticmethod
+    # def encode_token(obj):
+    #     jwt_payload_handler = api_settings.JWT_PAYLOAD_HANDLER
+    #     jwt_encode_handler = api_settings.JWT_ENCODE_HANDLER
+    #
+    #     payload = jwt_payload_handler(obj)
+    #     token = jwt_encode_handler(payload)
+    #     return token
+    @staticmethod
+    def encode_token(user, is_refresh_token=False):
```
Comment (mirr254) [inline authors/apps/authentication/backends.py]: I feel like the function name is not descriptive and a bit misleading. Kindly consider changing that to `generate_token`

File: authors/apps/authentication/backends.py [hunk 2]
```diff
@@ -1,10 +1,63 @@
...
+            'refresh_token': is_refresh_token,
+            'iat': datetime.datetime.utcnow(),
+            'nbf': datetime.datetime.utcnow() + datetime.timedelta(minutes=-5),
+            'exp': datetime.datetime.utcnow() + datetime.timedelta(days=7)
+        }, secret)
+        token = str(token)
+        logger.debug("is_refresh_token : %s : %s" % (is_refresh_token, token))
+        return token
 
+    # def generate_token(self, username, is_refresh_token=False):
```
Comment (mirr254) [inline authors/apps/authentication/backends.py]: Remove commented code

**Review by mirr254 — DISMISSED** [review]

File: authors/apps/authentication/backends.py (hunk 1 above)
Comment (EmmanuelBeja) [reply authors/apps/authentication/backends.py]: That makes sense. Am working on it.

File: authors/apps/authentication/backends.py (hunk 2 above)
Comment (EmmanuelBeja) [reply authors/apps/authentication/backends.py]: Am on it.

**Comment by GransonO** [comment]
Whats the progress on this so far?

**Comment by ghost** [comment]
@EmmanuelBeja , Good stuff, sir. I noticed there are merge conflicts though. Let's refine these so that this can be merged by time of demo

**Comment by ghost** [comment]
@EmmanuelBeja , Good stuff solving the conflict. 👍 

@mirr254 , this is now ready to merge

**Comment by GransonO** [comment]
I observe that the issues have been resolved and the PR meets the standards. Good work on this.
`LGTM`

**Review by ghost — APPROVED** [review]

**Review by mashafrancis — APPROVED** [review]
